### ISSUE TITLE
promql engine does not return expected results with mixed floats+histograms

### ISSUE THREAD

**Issue by jhesketh** [prometheus/prometheus#14172]
### What did you do?

I am testing the behavior (with promqltest) of querying a series that has both floats and histograms.

For example:

```
load 1m
	mixed_metric 1 2 3 {{schema:0 sum:5 count:4 buckets:[1 2 1]}} {{schema:0 sum:8 count:6 buckets:[1 4 1]}} 4 5 {{schema:0 sum:18 count:10 buckets:[3 4 3]}}

# I expect the following to work, but it Fails:
eval range from 0 to 8m step 1m mixed_metric
	{__name__="mixed_metric"} 1 2 3 {{count:4 sum:5 buckets:[1 2 1]}} {{count:6 sum:8 buckets:[1 4 1]}} 4 5 {{schema:0 sum:18 count:10 buckets:[3 4 3]}} {{schema:0 sum:18 count:10 buckets:[3 4 3]}}


# The following does pass though.
# Where we expected Histograms, we got Floats from a lookback
eval range from 0 to 8m step 1m mixed_metric
	{__name__="mixed_metric"} 1 2 3 3 3 4 5 {{schema:0 sum:18 count:10 buckets:[3 4 3]}} {{schema:0 sum:18 count:10 buckets:[3 4 3]}}


clear

# As another note, this only occurs when a float follows a histogram.
load 1m
	mixed_metric 1 2 3 {{schema:0 sum:5 count:4 buckets:[1 2 1]}} {{schema:0 sum:8 count:6 buckets:[1 4 1]}}

# Passes
eval range from 0 to 5m step 1m mixed_metric
	{__name__="mixed_metric"} 1 2 3 {{count:4 sum:5 buckets:[1 2 1]}} {{count:6 sum:8 buckets:[1 4 1]}} {{count:6 sum:8 buckets:[1 4 1]}}

# Fails
eval range from 0 to 5m step 1m mixed_metric
	{__name__="mixed_metric"} 1 2 3 3 3 {{count:6 sum:8 buckets:[1 4 1]}}

```

### What did you expect to see?

I expected to receive the Histogram point at 3m (in the first example).

### What did you see instead? Under which circumstances?

Instead, I received a Float of 3. Because another float exists later on in the series, a lookback was used instead of the Histogram that exists at that point.

**Comment by beorn7** [comment]
Thank you very much for the valuable bug report. 

@zenador I saw you already discussed this with @jhesketh .

Initial wild guess: We don't cut new chunks for floats, after there has been a histogram in between. And then the iterator just keeps going through the one float chunk.

**Comment by kushalShukla-web** [comment]
hey @jhesketh  `	{__name__="mixed_metric"} 1 2 3 {{count:4 sum:5 buckets:[1 2 1]}} {{count:6 sum:8 buckets:[1 4 1]}} 4 5 {{schema:0 sum:18 count:10 buckets:[3 4 3]}} {{schema:0 sum:18 count:10 buckets:[3 4 3]}}` here there is a extra histogram data at the end . that also because of the lookback?

**Comment by fatsheep9146** [comment]
> one

@beorn7 Just curious, is this a valid usage that prometheus should support? Or it's not a valid case that prometheus should panic or throw error in case this usage will not happen?

**Comment by beorn7** [comment]
Unfortunately, it can happen that a series switches between native histogram samples and float samples. It SHOULD NOT happen, but it can (e.g. if you switch a metric type without changing the name when updating an instrumented binary, or you manage to create the same metric from two different sources, commonly via recording rules). While users will rarely create this by accident, and they should generally avoid it, Prometheus has to handle the case gracefully.

**Comment by jhesketh** [comment]
> hey @jhesketh 	{__name__="mixed_metric"} 1 2 3 {{count:4 sum:5 buckets:[1 2 1]}} {{count:6 sum:8 buckets:[1 4 1]}} 4 5 {{schema:0 sum:18 count:10 buckets:[3 4 3]}} {{schema:0 sum:18 count:10 buckets:[3 4 3]}} here there is an extra histogram data at the end . that also because of the lookback?

(Sorry I missed this). Yes, that's a lookback because the selects 9 points. This is demonstrating that it correctly selected the last point which was a native-histogram rather than looking all the way back to 4.

**Comment by kushalShukla-web** [comment]
i liked to work on this issue @beorn7 .

**Comment by kushalShukla-web** [comment]
i would like to work on this issue .

**Comment by beorn7** [comment]
@kushalShukla-web I have assigned you to this issue.

Note that I suspect you have to go quite deep into the woods to fix this. You can always pull out if this feels too tough. Just let me know.

**Comment by kushalShukla-web** [comment]
Sure @beorn7

**Comment by kushalShukla-web** [comment]
I got this issue but I'm struggling to identify the exact location where the problem needs to be addressed.

**Comment by beorn7** [comment]
Finding the exact location is probably 70% of the solution here. This _should_ work, but somewhere in the code path, we forgot to check for this condition…

@kushalShukla-web you could see this as an opportunity to explore the (humongous) PromQL code base, with solving this issue as a byproduct. But it will be quite an investment. It's a totally acceptable decision to unassign yourself from this issue.

**Comment by kushalShukla-web** [comment]
i will give my 100% to solve this issue @beorn7 ,

**Comment by NeerajGartia21** [comment]
@beorn7 I would like to dig into it. As I'm working on #13934 , it would be easier for me!

**Comment by NeerajGartia21** [comment]
I think the issue is [here](https://github.com/prometheus/prometheus/blob/b6c538972c4d738896c0ee34fd7d8efcc6e2c2f8/tsdb/head_append.go#L1410), where we are committing the samples. For commiting the samples we are doing below process
```
	a.commitSamples(acc)
	a.commitHistograms(acc)
	a.commitFloatHistograms(acc)
	a.commitMetadata()
```
which commits the floats first and then commits the histograms, which in turn marks some of the histograms (which are there before floats) as Out-Of-Order and hence doesn't commits it to the `head`. I think here we can have a better commiting technique where the order of commits should be according the timestamps not according to the sample type.

**Comment by beorn7** [comment]
Interesting. I haven't checked in detail if that's actually an issue, but I remember that @krajorama talked about committing those samples in order for other reasons (cf. #15177).

**Comment by NeerajGartia21** [comment]
Yes, I checked that and the existing order of committing samples creates this type of issue when there is a mix of histograms and floats. So I think we should honor the order of samples.

**Comment by beorn7** [comment]
So appending everything in order will hit two birds with the same stone. @krajorama @NeerajGartia21 maybe you should coordinate your plans.

**Comment by beorn7** [comment]
We had a related [discussion on Slack](https://cloud-native.slack.com/archives/C02KR205UMU/p1730880687418859) (and yes, I fully acknowledge that it is bad to link something here that needs a Slack account to read, and as usual I want to state at this opportunity that open source projects SHOULD NOT use Slack as their means of communications). The bug we have at hand here can only happen if we mix floats and histograms in the same commit, so it will never happen during a regular scrape, but can only happen if ingesting via other means (remote-write receiver, OTLP receiver). Even then, there has to be a change from histograms to floats for the same series, which in itself is very rare (see above).
We should still solve this, but I would say it is not P1.

**Comment by alexgreenbank** [comment]
Hello from the bug scrub.

Work is ongoing.

**Comment by krajorama** [comment]
Hi,

I've started #16823 to deal with #15177. The test in this issue does pass in that PR, confirming @NeerajGartia21 's analysis.

**Comment by krajorama** [comment]
@beorn7 and I had a discussion about this problem and some questions that came up:

1. Beside unit tests when would this problem come up?
3. Could we introduce a contract to say that if the issue comes up , it is the users responsibility to use separate Appenders ?
4. Is there a way to efficiently detect if the problem comes up? If we could solve that, we could handle these - presumably - rare cases on a slow path and leave the current code alone.

Some other thoughts:
- The new `__type__` and `__unit__` labels don't solve the issue if it's between integer and float native histograms.
- Normally the problem wouldn't come up in scrape. But technically it's possible to encode a whole timeseries in exposition format using the explicit timestamp, so it could happen.
- Remote write protocol inherently has the same issue that it can have both float and histogram samples in the same time series, but we never merge on time, always process floats first. At least integer/float histograms are correctly handled.

**Comment by beorn7** [comment]
> Could we introduce a contract...

The contract would be: "All samples appended through this appender that belong to the same series must have the same type. It is the caller's responsibility to ensure that."

(The hope here is that the call sites that could potentially run into this problem already know about the structure of their data. If not, it could be quite expensive to detect.)

**Comment by ringerc** [comment]
I posted an analysis and report with test case here - https://github.com/prometheus/prometheus/issues/17025 - as I didn't find this issue when I hit problems writing my test cases for https://github.com/prometheus/prometheus/pull/17021 .

Regarding

> Beside unit tests when would this problem come up?

the only case I can think of would be a scrape that emitted both histogram and float samples in one scrape, with explicit timestamps. And I'm not sure if that's legal or not, it certainly wouldn't make a lot of sense for the workload to be doing that.

The main problem from my perspective is that this makes certain unit tests invisibly and confusingly produce wrong results. It took me most of a day to track it down - though I'm unsure how I didn't find this bug report.

**Reject mixed series in `promqltest`**

If `load` directives just rejected mixed series like this, where the 2nd one will trigger this bug:

```
load 10s
       data_histogram{type="mix_samples"} 0 1 {{schema:0 sum:1 count:2}} {{schema:0 sum:2 count:3}}
       data_histogram{type="mix_samples_hfirst"} {{schema:0 sum:1 count:2}} {{schema:0 sum:9 count:3}} 4 1
```

the problem would essentially go away.

Existing cases that currently mix histogram and float samples that happen to work because the histograms are newer than the floats can be updated to use split `load`s. This is a more realistic test anyway, since the only way you should be seeing mixed histogram and float samples in reality is from multiple scrapes.

E.g.: the above could be replaced with a split `load`.

```
load 10s
       data_histogram{type="mix_samples"} 0 1
       data_histogram{type="mix_samples_hfirst"} {{schema:0 sum:1 count:2}} {{schema:0 sum:9 count:3}}

load 10s
       data_histogram{type="mix_samples"} _ _  {{schema:0 sum:1 count:2}} {{schema:0 sum:2 count:3}}
       data_histogram{type="mix_samples_hfirst"} _ _ 4 1

eval instant at 1m count_over_time(data_histogram[2m])
       {type="mix_samples"} 4
       {type="mix_samples_hfirst"} 4
```

**Comment by ringerc** [comment]
It'd also be helpful for `load` directives to have a way to assert that various ingestion metrics like `headMetrics.outOfOrderSamples` (`prometheus_tsdb_out_of_order_samples_total`) has not increased during the `load`'s execution with an `expect` directive. This cannot be done with an `eval` because the TSDB metrics are not ingested during unit test execution.

The lack of any log output or any other way to observe that the metrics were dropped made this a _lot_ harder to track down.

The append contract sounds reasonable, though it'd be nice to have a way (even if slow and expensive) to enable checking of that contract for validating that a target or a unit test is doing the right thing.

**Comment by krajorama** [comment]
> the only case I can think of would be a scrape that emitted both histogram and float samples in one scrape, with explicit timestamps. And I'm not sure if that's legal or not, it certainly wouldn't make a lot of sense for the workload to be doing that.

Yes. Unfortunately that's true. I can imagine something batching together samples and exporting them as a batch. Not sure how it would work, but it's allowed for sure.

> Reject mixed series in promqltest

That seems like a good short term workaround to our problems with tests and the confusion they cause. Sorry about that!

Maybe that's the play here: reject such tests in the short term (prio P2). And work on a more long term solution (prio P3). Note: we'd like to make native histograms stable soon and get through P2 items.

**Comment by charleskorn** [comment]
> > Reject mixed series in promqltest
> 
> That seems like a good short term workaround to our problems with tests and the confusion they cause. Sorry about that!
> 
> Maybe that's the play here: reject such tests in the short term (prio P2). And work on a more long term solution (prio P3). Note: we'd like to make native histograms stable soon and get through P2 items.

If we implement the short-term solution here, how do you suggest we test scenarios with mixed floats and histograms? At present, it's possible to test at least some of these scenarios correctly provided the samples are in the right order, but this change would make that impossible.

**Comment by ringerc** [comment]
> If we implement the short-term solution here, how do you suggest we test scenarios with mixed floats and histograms? At present, it's possible to test at least some of these scenarios correctly provided the samples are in the right order, but this change would make that impossible.

I'd assumed that a PR that fixes the problem would remove the test restriction too.

Such a restriction can also be guarded by an internal-only feature flag (not exposed as an official feature flag) - so tests can set `promqltest.PermitMixedFloatHistogramLoads = true` or whatever, when they want to exercise the behaviour. Though thinking about it, that might not be too convenient to access with the test-script runner.

**Comment by beorn7** [comment]
Can we change the testing framework to add mixed samples in separate commits? Performance isn't really a problem when loading testdata.

**Comment by NeerajGartia21** [comment]
> Can we change the testing framework to add mixed samples in separate commits? Performance isn't really a problem when loading testdata.


I think we could handle this by using multiple appenders for series that contain different sample types. For example, commit one sample type first, then create a new appender to commit the other type. The downside is that performance isn’t ideal, and we can’t roll back partially committed samples if an error occurs.

**Comment by charleskorn** [comment]
> The downside is that performance isn’t ideal, and we can’t roll back partially committed samples if an error occurs.

I don't think either of these things matter much in unit tests.

**Comment by beorn7** [comment]
Indeed.

But maybe we'll solve this more fundamentally after all. I'll discuss things with @krajorama today in more depth. We'll report back in #15177.

**Comment by beorn7** [comment]
After our in-person discussion, we concluded we should bundle the discussion in #15177, which is essentially the root cause for the issue reported here.
